### PR TITLE
Fix bugs in primitives type validation and refactor

### PR DESCRIPTION
**Pull request by pedrorrivero**
### Summary
- Add type validation and type standardization to `BaseEstimator.run` to mitigate current bugs.
- Update `BaseEstimator._run` signature to expect `tuple`s instead of generic `Sequence`s.
- General refactor without extra features.

### Details and comments
This is not a breaking change, since any user implemented `BaseEstimator._run` ready to accept arbitrary `Sequence`s can accept `tuple`s as well.

### CODE REVIEW

**Comment by pedrorrivero** [comment]
Number types in `np.ndarray` need to be accounted for as well.

**Comment by pedrorrivero** [comment]
I cannot reproduce the failing test locally: `test.python.algorithms.minimum_eigensolvers.test_vqe.TestVQE.test_aux_operators_dict`

It passes on my machine under different versions of python, any idea @mtreinish ?

**Comment by pedrorrivero** [comment]
Even though I still cannot reproduce the failing test on my machine, I have traced back the possible causes.

It seems to come from ambiguity between a single empty element as input to `parameter_values` and the default for `None` when `len(circuits)==0`. If something like `parameter_values=[]` is provided, the current code assumes that this is a single element and parses it to `((),)`, even if it comes from the default expression `[()] * len(circuit)`.

The straightforward solution to this is to enforce `len(circuits)>0`, although for some reason this does not seem to fix failing test where an estimator is being called with zero input circuits. Notice how the error is always the same:

```
ValueError: The number of circuits (0) does not match the number of parameter value sets (1).
```

In my estimation, if this is the case, this is a bad use of the interface and what should be fixed is the test itself, allowed by a previous faulty validation on the primitives side.

---

I include a copy of the original tracebacks here for easy reference:
```
==============================
Failed 4 tests - output below:
==============================

test.python.algorithms.minimum_eigensolvers.test_vqe.TestVQE.test_aux_operators_dict
------------------------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):

      File "/home/vsts/work/1/s/qiskit/algorithms/observables_evaluator.py", line 68, in estimate_observables
    estimator_job = estimator.run(quantum_state, observables_list, parameter_values)

      File "/home/vsts/work/1/s/qiskit/primitives/base_estimator.py", line 524, in run
    f"The number of circuits ({len(circuits)}) does not match "

    ValueError: The number of circuits (0) does not match the number of parameter value sets (1).

    
The above exception was the direct cause of the following exception:


    Traceback (most recent call last):

      File "/tmp/terra-tests/test/python/algorithms/minimum_eigensolvers/test_vqe.py", line 407, in test_aux_operators_dict
    result = vqe.compute_minimum_eigenvalue(self.h2_op, aux_operators={})

      File "/home/vsts/work/1/s/qiskit/algorithms/minimum_eigensolvers/vqe.py", line 198, in compute_minimum_eigenvalue
    self.estimator, self.ansatz, aux_operators, optimizer_result.x

      File "/home/vsts/work/1/s/qiskit/algorithms/observables_evaluator.py", line 71, in estimate_observables
    raise AlgorithmError("The primitive job failed!") from exc

    qiskit.algorithms.exceptions.AlgorithmError: 'The primitive job failed!'


test.python.algorithms.minimum_eigensolvers.test_vqe.TestVQE.test_aux_operators_list
------------------------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):

      File "/home/vsts/work/1/s/qiskit/algorithms/observables_evaluator.py", line 68, in estimate_observables
    estimator_job = estimator.run(quantum_state, observables_list, parameter_values)

      File "/home/vsts/work/1/s/qiskit/primitives/base_estimator.py", line 524, in run
    f"The number of circuits ({len(circuits)}) does not match "

    ValueError: The number of circuits (0) does not match the number of parameter value sets (1).

    
The above exception was the direct cause of the following exception:


    Traceback (most recent call last):

      File "/tmp/terra-tests/test/python/algorithms/minimum_eigensolvers/test_vqe.py", line 369, in test_aux_operators_list
    result = vqe.compute_minimum_eigenvalue(self.h2_op, aux_operators=[])

      File "/home/vsts/work/1/s/qiskit/algorithms/minimum_eigensolvers/vqe.py", line 198, in compute_minimum_eigenvalue
    self.estimator, self.ansatz, aux_operators, optimizer_result.x

      File "/home/vsts/work/1/s/qiskit/algorithms/observables_evaluator.py", line 71, in estimate_observables
    raise AlgorithmError("The primitive job failed!") from exc

    qiskit.algorithms.exceptions.AlgorithmError: 'The primitive job failed!'


test.python.algorithms.test_observables_evaluator.TestObservablesEvaluator.test_estimate_observables_5___
---------------------------------------------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):

      File "/home/vsts/work/1/s/qiskit/algorithms/observables_evaluator.py", line 68, in estimate_observables
    estimator_job = estimator.run(quantum_state, observables_list, parameter_values)

      File "/home/vsts/work/1/s/qiskit/primitives/base_estimator.py", line 524, in run
    f"The number of circuits ({len(circuits)}) does not match "

    ValueError: The number of circuits (0) does not match the number of parameter value sets (1).

    
The above exception was the direct cause of the following exception:


    Traceback (most recent call last):

      File "/home/vsts/work/1/s/test-job/lib/python3.7/site-packages/ddt.py", line 220, in wrapper
    return func(self, *args, **kwargs)

      File "/tmp/terra-tests/test/python/algorithms/test_observables_evaluator.py", line 132, in test_estimate_observables
    estimator,

      File "/tmp/terra-tests/test/python/algorithms/test_observables_evaluator.py", line 75, in _run_test
    result = estimate_observables(estimator, quantum_state, observables, None, self.threshold)

      File "/home/vsts/work/1/s/qiskit/algorithms/observables_evaluator.py", line 71, in estimate_observables
    raise AlgorithmError("The primitive job failed!") from exc

    qiskit.algorithms.exceptions.AlgorithmError: 'The primitive job failed!'


test.python.algorithms.test_observables_evaluator.TestObservablesEvaluator.test_estimate_observables_6
------------------------------------------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):

      File "/home/vsts/work/1/s/qiskit/algorithms/observables_evaluator.py", line 68, in estimate_observables
    estimator_job = estimator.run(quantum_state, observables_list, parameter_values)

      File "/home/vsts/work/1/s/qiskit/primitives/base_estimator.py", line 524, in run
    f"The number of circuits ({len(circuits)}) does not match "

    ValueError: The number of circuits (0) does not match the number of parameter value sets (1).

    
The above exception was the direct cause of the following exception:


    Traceback (most recent call last):

      File "/home/vsts/work/1/s/test-job/lib/python3.7/site-packages/ddt.py", line 220, in wrapper
    return func(self, *args, **kwargs)

      File "/tmp/terra-tests/test/python/algorithms/test_observables_evaluator.py", line 132, in test_estimate_observables
    estimator,

      File "/tmp/terra-tests/test/python/algorithms/test_observables_evaluator.py", line 75, in _run_test
    result = estimate_observables(estimator, quantum_state, observables, None, self.threshold)

      File "/home/vsts/work/1/s/qiskit/algorithms/observables_evaluator.py", line 71, in estimate_observables
    raise AlgorithmError("The primitive job failed!") from exc

    qiskit.algorithms.exceptions.AlgorithmError: 'The primitive job failed!'
```

**Comment by pedrorrivero** [comment]
It looks like it is confirmed: the failing tests are indeed calling the estimator with no circuits. After disallowing that explicitly, this is the new error they raise:
```
ValueError: No circuits were provided.
```

**Comment by mrossinek** [comment]
@pedrorrivero Thanks for the ping!

I believe this is a short-coming of the `estimate_observables` method, which should guard against an empty set of observables being provided to it. The reason that you also see `VQE` tests failing, is that it uses that method internally, and currently also only guards against an `is None` check.

How would you like to proceed here? Shall I open a separate PR to fix the short-comings of the `estimate_observables` method?

**Comment by ElePT** [comment]
Hey @pedrorrivero! I agree that the `Estimator` should handle empty lists of observables anyways, but I have opened a PR to fix the `estimate_observables` method not to send empty lists (after all, it would be adding unnecessary I/O overhead to just get an empty list in return). Thanks for the catch :)

File: qiskit/primitives/base_estimator.py [hunk 1]
```diff
@@ -170,6 +171,212 @@ def __init__(
...
+        self._run_options.update_options(**fields)
+
+    ################################################################################
+    ## METHODS
+    ################################################################################
+    def run(
+        self,
+        circuits: Sequence[QuantumCircuit] | QuantumCircuit,
+        observables: Sequence[BaseOperator | PauliSumOp] | BaseOperator | PauliSumOp,
+        parameter_values: Sequence[Sequence[float | int]] | Sequence[float | int] | None = None,
```
[the remark this answers is not in the document]
Comment (mrossinek) [reply qiskit/primitives/base_estimator.py:201]: Huh... `mypy` treats them as implied but I'm guessing that is another inconsistency in the numeric stack of Python type-handling then :confused:

**Comment by pedrorrivero** [comment]
Thanks for the quick response! You are awesome!

I think a separate PR would be best as @ElePT did, @mrossinek, I don't know if you have talked over that 🙂 

Also @ElePT, this new implementation of `BaseEstimator` DOES allow to get empty `observables` but it raises a `ValueError` because it needs at least one to do the computation 😅 Is that what you mean?

```python
    @staticmethod
    def _validate_observables(
        observables: Sequence[BaseOperator | PauliSumOp] | BaseOperator | PauliSumOp,
    ) -> tuple[BaseOperator | PauliSumOp, ...]:
        if isinstance(observables, (BaseOperator, PauliSumOp)):
            observables = (observables,)
        elif not isinstance(observables, Sequence) or not all(
            isinstance(obs, (BaseOperator, PauliSumOp)) for obs in observables
        ):
            raise TypeError("Invalid observables, expected Sequence[BaseOperator|PauliSumOp].")
        elif not isinstance(observables, tuple):
            observables = tuple(observables)
        if len(observables) == 0:
            raise ValueError("No observables were provided.")
        return observables

```

File: qiskit/primitives/base_estimator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (woodsp-ibm) [reply qiskit/primitives/base_estimator.py:201]: https://peps.python.org/pep-0484/#the-numeric-tower

[the remark this answers is not in the document]
Comment (pedrorrivero) [reply qiskit/primitives/base_estimator.py]: I have updated the functionality then to account for this. Thanks @woodsp-ibm and @mrossinek for noticing this!

File: qiskit/primitives/base_estimator.py [hunk 2]
```diff
@@ -170,6 +171,212 @@ def __init__(
...
+            **run_opts.__dict__,
+        )
+
+    # This will be comment out after 0.22. (This is necessary for the compatibility.)
+    # @abstractmethod
+    def _run(
+        self,
+        circuits: tuple[QuantumCircuit, ...],
+        observables: tuple[BaseOperator | PauliSumOp, ...],
+        parameter_values: tuple[tuple[float, ...], ...],
```
[the remark this answers is not in the document]
Comment (ikkoham) [reply qiskit/primitives/base_estimator.py:288]: `np.ndarray` is not `Sequence`. You can register it like `Sequence.register(np.ndarray)`.
or should define `Protocol` if you want.

If you need to use some tuple feature, you could use tuple for it, but I don't want to convert data as much as possible. There is an overhead.

[the remark this answers is not in the document]
Comment (pedrorrivero) [reply qiskit/primitives/base_estimator.py:290]: Thanks for the `Sequence.register(np.ndarray)` trick @ikkoham ! I had it on my to-dos, I have already taken care of it 😄

**Comment by pedrorrivero** [comment]
After meeting we have agreed on disallowing empty inputs (i.e. throw error). This is already included in this PR.

**Comment by ikkoham** [comment]
(offline discussion) I agreed with using `tuple`.

**Comment by pedrorrivero** [comment]
Code repetition in `BaseSampler` between `BaseEstimator` is avoided through a new shared parent class: `BasePrimitive`

**Comment by pedrorrivero** [comment]
Tests are failing due to `retworkx`, must come from latest marge of `main` @mtreinish

File: qiskit/primitives/base/base_estimator.py [hunk 3]
```diff
@@ -90,16 +92,18 @@
...
 from qiskit.quantum_info.operators import SparsePauliOp
 from qiskit.quantum_info.operators.base_operator import BaseOperator
 from qiskit.utils.deprecation import deprecate_arguments, deprecate_function
 
+from .base_primitive import BasePrimitive
 from .estimator_result import EstimatorResult
-from .utils import _circuit_key, _observable_key, init_observable
+from ..utils import _circuit_key, _observable_key, init_observable
 
+Sequence.register(np.ndarray)
```
Comment (jakelishman) [inline qiskit/primitives/base/base_estimator.py:103]: You can't do this - Numpy would do it if it were valid. `ndarray` doesn't fulfil all the methods of `Sequence` (e.g. `index`) so it isn't one. I haven't read the PR fully, but on a quick scan this seems to be used as a type check only, before you normalise to `tuple`. If you're normalising it to a tuple, you don't need something to be `Sequence`, but generally too, this sort of super-precise type checking isn't necessary.  You've already got the type hinting and documentation, and the `tuple` calls will also give a sensible error message if something fails. Overdoing this just causes slowdown for every user who _is_ doing the right thing. As you're finding here, it also means you originally cut out a bunch of valid inputs, and who's to say it's only `ndarray`?

In general, we also should not be affecting global state like this for any objects we don't own. It's not our decision to make for users.

**Review by jakelishman — CHANGES_REQUESTED** [review]

File: qiskit/primitives/base/base_estimator.py (hunk 3 above)
Comment (pedrorrivero) [reply qiskit/primitives/base/base_estimator.py]: Thanks so much @jakelishman , I was concerned about `Sequence.register` but under the impression that we had to do it. Glad that you stepped in, it is already fixed 🙂 

Regarding parsing the inputs, we had internal conversations about it and your same concerns were raised. I will fill you in on the details on why we decided to go for this separately, and see what your opinions are.

**Comment by pedrorrivero** [comment]
I have reviewed the entire PR and removed all the potential conflicts with Qiskit Terra `0.22.0rc1`, so that it can be *merged in time for the `0.22.0` official release*.

Sadly, not going for the full PR means that we will probably have to introduce new deprecations in `0.23.0`. We can deal with this separately.

@ikkoham @t-imamichi

File: qiskit/primitives/base/base_estimator.py (hunk 3 above)
Comment (jakelishman) [reply qiskit/primitives/base/base_estimator.py]: It's ok, my stylistic opinions aren't generally relevant here, since I'm not a principal maintainer of these sections. I only commented at all because there was a technical issue I'd seen.

**Review by t-imamichi — DISMISSED** [review]
Thank you for adding extensive input checks for primitives. LGTM.

**Comment by pedrorrivero** [comment]
Solved conflict from #8860

**Review by ikkoham — APPROVED** [review]
Thank you for adding the validations and refactoring. LGTM
